vertx-web-graphql/src/main/java/io/vertx/ext/web/handler/graphql/impl/ApolloWSHandlerImpl.java

```diff
@@ -0,0 +1,243 @@
+import io.vertx.ext.web.RoutingContext;
+import io.vertx.ext.web.handler.graphql.ApolloWSHandler;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * @author Rogelio Orts
+ */
+public class ApolloWSHandlerImpl implements ApolloWSHandler {
+
+  private final static String HEADER_CONNECTION_UPGRADE_VALUE = "upgrade";
+
+  private final GraphQL graphQL;
+
+  private Handler<ServerWebSocket> endHandler;
+
+  public ApolloWSHandlerImpl(GraphQL graphQL) {
+    this.graphQL = graphQL;
+  }
+
+  @Override
+  public void handle(RoutingContext routingContext) {
+    MultiMap headers = routingContext.request().headers();
+    if(
+      headers.contains(HttpHeaders.CONNECTION)
+        &&
+      HEADER_CONNECTION_UPGRADE_VALUE.equals(headers.get(HttpHeaders.CONNECTION).toLowerCase())
+    ) {
+      ServerWebSocket serverWebSocket = routingContext.request().upgrade();
+      handleConnection(serverWebSocket);
+    } else {
+      routingContext.next();
+    }
+  }
+
+  private void handleConnection(ServerWebSocket serverWebSocket) {
+    final Map<String, Subscription> subscriptions = Collections.synchronizedMap(new HashMap<>());
+
+    serverWebSocket.handler(buffer -> {
+      try {
+        JsonObject message = buffer.toJsonObject();
+        String opId = message.getString("id");
+        String type = message.getString("type");
+
+        if (type == null) {
+          sendError(serverWebSocket, opId, new Exception("Invalid message type!"));
+          return;
+        }
+
+        switch (ApolloWSMessageType.from(type)) {
+          case CONNECTION_INIT:
+            sendMessage(serverWebSocket, null, ApolloWSMessageType.CONNECTION_ACK);
+            break;
+          case CONNECTION_TERMINATE:
+            serverWebSocket.close();
+            break;
+          case START:
+            start(serverWebSocket, subscriptions, message);
+            break;
+          case STOP:
+            stop(serverWebSocket, subscriptions, opId);
+            break;
+          default:
+            sendError(serverWebSocket, opId, new Exception("Invalid message type!"));
+            break;
+        }
+      } catch (Exception e) {
+        sendError(serverWebSocket, null, e);
+      }
+    });
+
+    serverWebSocket.endHandler(v -> {
+      subscriptions.values().forEach(Subscription::cancel);
+
+      if (endHandler != null) {
+        endHandler.handle(serverWebSocket);
+      }
+    });
+  }
+
+  @Override
+  public ApolloWSHandler endHandler(Handler<ServerWebSocket> endHandler) {
+    this.endHandler = endHandler;
+
+    return this;
+  }
+
+  private void start(
+    ServerWebSocket serverWebSocket,
+    Map<String, Subscription> subscriptions,
+    JsonObject message
+  ) {
+    String opId = message.getString("id");
+
+    // Unsubscribe if it's subscribed
+    if (subscriptions.containsKey(opId)) {
```

Comment:
How could this happen?

Reply:
By sending a start message with an opId and then sending again the a start message with the same opId. Maybe by error or whatever, but it'd receive duplicate data if we don't unsubscribe it before.

Reply:
Ok, thanks for the details. So that would be if the Apollo client gets buggy or a user writes a custom client.